tsweb/varz/varz.go

```diff
@@ -316,21 +317,24 @@ type PrometheusMetricsReflectRooter interface {
 
 var expvarDo = expvar.Do // pulled out for tests
 
-func writeMemstats(w io.Writer, ms *runtime.MemStats) {
-	out := func(name, typ string, v uint64, help string) {
-		if help != "" {
-			fmt.Fprintf(w, "# HELP memstats_%s %s\n", name, help)
-		}
-		fmt.Fprintf(w, "# TYPE memstats_%s %s\nmemstats_%s %v\n", name, typ, name, v)
+func printfMemstats[V constraints.Integer | constraints.Float](w io.Writer, name, typ string, v V, help string) {
+	if help != "" {
+		w.Write([]byte("# HELP memstats_" + name + " " + help + "\n"))
 	}
-	g := func(name string, v uint64, help string) { out(name, "gauge", v, help) }
-	c := func(name string, v uint64, help string) { out(name, "counter", v, help) }
-	g("heap_alloc", ms.HeapAlloc, "current bytes of allocated heap objects (up/down smoothly)")
-	c("total_alloc", ms.TotalAlloc, "cumulative bytes allocated for heap objects")
-	g("sys", ms.Sys, "total bytes of memory obtained from the OS")
-	c("mallocs", ms.Mallocs, "cumulative count of heap objects allocated")
-	c("frees", ms.Frees, "cumulative count of heap objects freed")
-	c("num_gc", uint64(ms.NumGC), "number of completed GC cycles")
+	w.Write([]byte(
+		"# TYPE memstats_" + name + " " + typ + "\n" +
+			"memstats_" + name + " " + fmt.Sprintf("%v", v) + "\n",
+	))
```

Comment:
No, let's not do this. This is gross and needlessly alloc-y and still uses Sprintf after all that.